Snowflake.Data.Tests/UnitTests/SFGCSClientTest.cs

```diff
@@ -0,0 +1,457 @@
+        public async Task TestUploadFileAsync(HttpStatusCode httpStatusCode, ResultStatus expectedResultStatus)
+        {
+            // Arrange
+            var mockWebRequest = new Mock<WebRequest>();
+            mockWebRequest.Setup(c => c.Headers).Returns(new WebHeaderCollection());
+            mockWebRequest.Setup(client => client.GetResponseAsync())
+                .Returns(() =>
+                {
+                    var response = new Mock<HttpWebResponse>();
+
+                    if (httpStatusCode != HttpStatusCode.OK)
+                    {
+                        response.SetupGet(c => c.StatusCode)
+                            .Returns(httpStatusCode);
+                        throw new WebException("Mock GCS Error", null, 0, response.Object);
+                    }
+
+                    return Task.FromResult((WebResponse)response.Object);
+                });
+            mockWebRequest.Setup(client => client.GetRequestStreamAsync())
+                .Returns(() => Task.FromResult((Stream) new MemoryStream()));
+            _client.SetCustomWebRequest(mockWebRequest.Object);
+            _fileMetadata.uploadSize = UploadFileSize;
+
+            // Act
+            await _client.UploadFileAsync(_fileMetadata, new MemoryStream(), new SFEncryptionMetadata()
+            {
+                iv = MockGCSClient.GcsIV,
+                key = MockGCSClient.GcsKey,
+                matDesc = MockGCSClient.GcsMatdesc
+            },
+            _cancellationToken).ConfigureAwait(false);
+
+            // Assert
+            AssertForUploadFileTests(expectedResultStatus);
+        }
+
+        private void AssertForUploadFileTests(ResultStatus expectedResultStatus)
+        {
+            if (expectedResultStatus == ResultStatus.UPLOADED)
+            {
+                Assert.AreEqual(_fileMetadata.uploadSize, _fileMetadata.destFileSize);
+            }
+
+            Assert.AreEqual(expectedResultStatus.ToString(), _fileMetadata.resultStatus);
+        }
+
+        [Test]
+        [TestCase(HttpStatusCode.OK, ResultStatus.DOWNLOADED)]
+        [TestCase(HttpStatusCode.Unauthorized, ResultStatus.RENEW_TOKEN)]
+        [TestCase(HttpStatusCode.Forbidden, ResultStatus.NEED_RETRY)]
+        [TestCase(HttpStatusCode.InternalServerError, ResultStatus.NEED_RETRY)]
+        [TestCase(HttpStatusCode.ServiceUnavailable, ResultStatus.NEED_RETRY)]
+        public void TestDownloadFile(HttpStatusCode httpStatusCode, ResultStatus expectedResultStatus)
+        {
+            // Arrange
+            var mockWebRequest = new Mock<WebRequest>();
+            mockWebRequest.Setup(client => client.GetResponse())
+                .Returns(() =>
+                {
+                    var response = new Mock<HttpWebResponse>();
+
+                    if (httpStatusCode == HttpStatusCode.OK)
+                    {
+                        response.Setup(c => c.Headers).Returns(new WebHeaderCollection());
```

Comment:
this response.Object is pretty much the same in Sync and Async tests and could be extracted - diff is about wrapping it in Task. I'm wondering if it would make this test more clear

Reply:
I've put it in a method